File: clients/mobile/lib/utils/purchase_order_util.dart
/// PURCHASE_ORDER receive + line helpers — web `order-line` / entity-record parity.

const purchaseOrderLineEntityCode = 'PURCHASE_ORDER_LINE';
const purchaseOrderParentFkField = 'po_id';
const purchaseOrderCreatePrefillParam = 'po_id';

bool canReceivePurchaseOrderStatus(Map<String, dynamic> record) {
  final status = '${record['status'] ?? ''}';
  return status == 'draft' || status == 'submitted';
}

bool canReceivePurchaseOrder(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
  int orderLineCount = 0,
}) {
  final id = recordId ?? '${record['id'] ?? ''}';
  if (entityCode != 'PURCHASE_ORDER' || id.isEmpty || creatingNew) {
    return false;
  }
  if (orderLineCount <= 0) {
    return false;
  }
  return canReceivePurchaseOrderStatus(record);
}

bool canAddPurchaseOrderLine(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
}) {
  final id = recordId ?? '${record['id'] ?? ''}';
  if (entityCode != 'PURCHASE_ORDER' || id.isEmpty || creatingNew) {
    return false;
  }
  return canReceivePurchaseOrderStatus(record);
}

List<Map<String, dynamic>> filterPurchaseOrderLines(
  List<Map<String, dynamic>> lines,
  String poId,
) {
  return lines
      .where((row) => '${row[purchaseOrderParentFkField] ?? ''}' == poId)
      .toList();
}

double purchaseOrderLineQuantity(Map<String, dynamic> line) {
  final value = double.tryParse('${line['quantity'] ?? ''}');
  return value ?? 0;
}

double purchaseOrderLineUnitPrice(Map<String, dynamic> line) {
  final value = double.tryParse('${line['unit_price'] ?? ''}');
  return value ?? 0;
}

double purchaseOrderLineExtension(Map<String, dynamic> line) {
  return purchaseOrderLineQuantity(line) * purchaseOrderLineUnitPrice(line);
}

File: clients/web/src/app/shared/utils/purchase_order_util.ts
import {
  filterOrderLines,
  PO_LINE_PARENT_FIELD,
} from './order-line.util';

export const PO_LINE_ENTITY_CODE = 'PURCHASE_ORDER_LINE';
export const PO_CREATE_PREFILL_PARAM = 'po_id';

/** Whether PO status allows receive or line edits — mobile `canReceivePurchaseOrder` parity. */
export function canReceivePurchaseOrderStatus(record: Record<string, unknown>): boolean {
  const status = String(record['status'] ?? '');
  return status === 'draft' || status === 'submitted';
}

export function canReceivePurchaseOrder(
  entityCode: string,
  record: Record<string, unknown>,
  options: {
    recordId?: string | null;
    creatingNew?: boolean;
    orderLineCount?: number;
  } = {},
): boolean {
  const { recordId, creatingNew = false, orderLineCount = 0 } = options;
  const id = recordId ?? String(record['id'] ?? '');
  if (entityCode !== 'PURCHASE_ORDER' || !id || creatingNew) {
    return false;
  }
  if (orderLineCount <= 0) {
    return false;
  }
  return canReceivePurchaseOrderStatus(record);
}

export function canAddPurchaseOrderLine(
  entityCode: string,
  record: Record<string, unknown>,
  options: { recordId?: string | null; creatingNew?: boolean } = {},
): boolean {
  const { recordId, creatingNew = false } = options;
  const id = recordId ?? String(record['id'] ?? '');
  if (entityCode !== 'PURCHASE_ORDER' || !id || creatingNew) {
    return false;
  }
  return canReceivePurchaseOrderStatus(record);
}

export function filterPurchaseOrderLines(
  lines: Record<string, unknown>[],
  poId: string,
): Record<string, unknown>[] {
  return filterOrderLines(lines, PO_LINE_PARENT_FIELD, poId);
}
